[Bug 16407] Splitting by empty should be an error

peter-b opened the pull request:
`MCStringSplit()` misbehaves when it receives an empty delimiter
string, entering an infinite loop.

~~This patch alters the behaviour of `MCStringSplit()` to treat empty
delimiters as never existing in the target string.  This is consistent
with the behaviour of the `contains` operator:
`"abc" contains empty -> false`~~

This patch turns splitting by an empty delimiter into an error, for consistency with LiveCode 6.7.  It also fixes an inconsistency between the Unicode and native character implementations of splitting.

runrevmark commented:
The behavior of split by empty should be to throw an error as per 6.7 - at the level of script at least. (I.e. in ExecSplit). The problem was that when split was refactored the check for delimiter being 1 char long was removed, rather than being replaced with one for >= 1.

In terms of 'doing something sensible' if you pass in an empty delimiter to the libfoundation function - then I think having it work the same as 'not found' makes perfect sense, however.

peter-b commented:
But _why_ does `split by empty` throw an error but `contains empty` doesn't?

runrevmark commented:
I couldn't really say - for the develop-7.0 patch it is probably better to put the check at the LCS level (reinstating what was there in 6.7 - but obviously allowing delimiters >= 1 in length). For 8.0, however, it's fine to expose the better behavior which is consistent with contains - the docs for split should be updated to reflect that.

peter-b commented:
Okay -- I'll refactor the two pull requests to do that (and to make sure that the "right" result ends up in the develop branch).

peter-b committed:
[Bug 16407] Splitting by empty should be an error.

Modify the engine to make attempting to split a string with an empty
delimiter into an error.  This is to fix a regression relative to the
6.7 behaviour, and to prevent an infinite loop in `MCStringSplit`.

peter-b committed:
libfoundation: Fix Unicode string split with missing key delimiter

When splitting Unicode strings, make the behaviour when the key
delimiter is not found consistent with the behaviour for native
strings.

peter-b commented:
@runrevmark Is this better?

runrevmark commented on engine/src/cmdsm.cpp:
```diff
@@ -1327,6 +1327,14 @@ void MCArrayOp::exec_ctxt(MCExecContext &ctxt)
 	else
 	{
+		/* Delimiters cannot be empty for split */
+		if (MCStringIsEmpty(*t_element_del) ||
```
This check needs to go into the ExecSplit\* methods. This is a runtime check after type-conversion - the code in the exec_ctxt methods should be (as far as is possible) type-convert, select method, dispatch.

peter-b committed:
[Bug 16407] Move split by empty check into Exec methods

peter-b commented:
@runrevmark Okay, I've moved the checks into the Exec methods and added the appropriate tests.

runrevmark commented:
@livecode-vulcan review ok 9641b76

livecode-vulcan commented:
💙 review by @runrevmark ok 9641b7627aaed21b333320bc9ca005127983a2cf

livecode-vulcan commented:
😎 test success 9641b7627aaed21b333320bc9ca005127983a2cf
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1157)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1168)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1154)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1160)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1172)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1174)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1176)
